File: app/src/main/java/com/termux/utils/Extrakeys.java
package com.termux.utils;

import android.graphics.Canvas;
import android.view.KeyEvent;
import android.view.MotionEvent;
import android.view.View;

import com.termux.NyxActivity;
import com.termux.data.Properties;

import java.util.ArrayList;

import nyx.constants.Constant;

public final class Extrakeys extends View {
    private static final int buttonRadius = 30;
    private final float offsetText = UiElements.paint.descent();
    private final ArrayList<Key> keys = new ArrayList<>(5);
    private float a;

    public Extrakeys() {
        super(NyxActivity.console.getContext());
        this.setFocusable(false);
        this.keys.add(new Key("⇪", 0));
        this.keys.add(new Key("C", 0));
        this.keys.add(new Key("A", 0));
        this.keys.add(new Key("Fn", 0));
        new Properties(Constant.EXTRA_KEYS_CONFIG).forEach((it, value) -> this.keys.add(new Key(value, Integer.parseInt(it))));
    }

    @Override
    protected void onSizeChanged(final int w, final int h, final int oldw, final int oldh) {
        if (0 == w || 0 == h) return;
        final var centerX = w / 2;
        final var numButtons = this.keys.size();
        final float angle = (float) Math.asin(Extrakeys.buttonRadius * 2.0f / centerX) + 0.07f;
        var centeringOffset = 3.14f / 2 + angle * (numButtons - 1) / 2;
        this.a = (centerX - (Extrakeys.buttonRadius + 5));
        for (final Key i : this.keys) {
            i.x = (float) (centerX + this.a * Math.cos(centeringOffset));
            i.y = (float) (centerX + this.a * Math.sin(centeringOffset));
            centeringOffset -= angle;
        }
    }

    @Override
    protected void onDraw(final Canvas canvas) {
        int n = 0;
        for (final Key k : this.keys) {
            UiElements.paint.setColor((4 > n && NyxActivity.console.metaKeys[n]) ? Theme.primary : Theme.secondary);
            canvas.drawCircle(k.x, k.y, Extrakeys.buttonRadius, UiElements.paint);
            UiElements.paint.setColor(Theme.getContrastColor(UiElements.paint.getColor()));
            canvas.drawText(k.label, k.x, k.y + this.offsetText, UiElements.paint);
            n++;
        }
    }

    @Override
    public boolean onTouchEvent(final MotionEvent event) {
        final var x = event.getX();
        final var y = event.getY();
        if (MotionEvent.ACTION_DOWN == event.getAction() && !UiElements.inCircle(this.getWidth() / 2.0f, this.getHeight() / 2.0f, (this.a - Extrakeys.buttonRadius), x, y)) {
            int n = 0;
            for (final Key i : this.keys) {
                if (UiElements.inCircle(i.x, i.y, Extrakeys.buttonRadius, x, y)) {
                    if (4 > n) {
                        NyxActivity.console.metaKeys[n] = !NyxActivity.console.metaKeys[n];
                        this.invalidate();
                    } else
                        NyxActivity.console.dispatchKeyEvent(new KeyEvent(KeyEvent.ACTION_DOWN, i.code));
                    return true;
                }
                n++;
            }
        }
        return false;
    }

    private static class Key {
        final int code;
        final String label;
        float x;
        float y;

        Key(final String l, final int c) {
            this.label = l;
            this.code = c;
        }
    }

}

File: app/src/main/java/com/termux/utils/ui/Extrakeys.kt
package com.termux.utils.ui

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.view.KeyEvent
import android.view.MotionEvent
import android.view.View
import com.termux.utils.data.ConfigManager
import com.termux.utils.data.ConfigManager.CONFIG_PATH
import com.termux.utils.data.Properties
import com.termux.utils.data.TerminalManager.console
import com.termux.utils.data.isPointInCircle
import kotlin.math.asin
import kotlin.math.cos
import kotlin.math.sin

private class Key(val label: String, val code: Int)

private const val buttonRadius = 25f

class Extrakeys(context: Context) : View(context) {
    private var a = 0f
    val paint: Paint = Paint().apply {
        typeface = ConfigManager.typeface
        textSize = 35f
        textAlign = Paint.Align.CENTER
        color = colorPrimaryAccent
    }
    private val buttonStateRefs = arrayOf(
        console::isControlKeydown,
        console::isReadAltKey,
        console::isReadShiftKey,
        console::readFnKey
    )
    private val offsetText = paint.descent()
    private val posMap = mutableMapOf<Float, Float>()

    private val normalKey = mutableListOf<Key>()

    init {
        val properties = Properties("$CONFIG_PATH/keys")
        properties.forEach { it, value ->
            normalKey.add(Key(it, value.toInt()))
        }

        isFocusable = false
    }


    private val numButtons = buttonStateRefs.size + normalKey.size - 1
    private val label = arrayOf("C", "A", "S", "Fn")

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        posMap.clear()
        val centerX = w / 2
        val angle = asin(buttonRadius * 2 / centerX) + (3.14f / 45)
        var centeringOffset = 3.14f / 2 + angle * numButtons / 2
        a = (centerX - (buttonRadius + 5))
        for (i in 0..numButtons) {
            posMap[(centerX + a * cos(centeringOffset))] = centerX + a * sin(centeringOffset)
            centeringOffset -= angle
        }
    }


    override fun onDraw(canvas: Canvas) {
        var n = 0
        for ((key, value) in posMap) {
            paint.color =
                if (n < buttonStateRefs.size && buttonStateRefs[n].get()) colorPrimaryAccent else surface
            canvas.drawCircle(
                key, value, buttonRadius, paint
            )
            paint.color = primaryTextColor
            val text =
                if (n < buttonStateRefs.size) label[n] else normalKey[n - buttonStateRefs.size].label
            canvas.drawText(text, key, value + offsetText, paint)
            n++
        }
        super.onDraw(canvas)
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        val x = event.x
        val y = event.y
        var n = 0
        if (event.action == MotionEvent.ACTION_DOWN && !isPointInCircle(
                width / 2f, height / 2f, (a - buttonRadius), x, y
            )
        ) {
            for ((key, value) in posMap) {
                if (isPointInCircle(
                        key, value, buttonRadius, x, y
                    )
                ) {
                    if (n < buttonStateRefs.size) buttonStateRefs[n].set(!buttonStateRefs[n].get())
                    else console.dispatchKeyEvent(
                        KeyEvent(
                            KeyEvent.ACTION_DOWN, normalKey[n - buttonStateRefs.size].code
                        )
                    )
                    invalidate()
                    return true
                }
                n++
            }
        }
        return false
    }
}
